image_captioning/models/decoder.py

```diff
@@ -0,0 +1,120 @@
+"""
+Decoder model.
+"""
+from typing import Tuple
+
+import tensorflow as tf
+
+from .attention import BahdanauAttention
+
+
+class Decoder(tf.keras.Model):
+    """Decoder model.
+    """
+
+    def __init__(self, hidden_units: int, attn_hidden_units: int,
+                 vocab_size: int, embedding_dim: int):
+        """
+        Arguments:
+            hidden_units: Number of hidden units for LSTM.
+            attn_hidden_units: Number of hidden units for attention module.
+            vocab_size: Number of possible vocabularies in the language.
+            embedding_dim: Dimension of word embedding vector.
+        """
+        super().__init__(self)
+
+        self.hidden_units = hidden_units
+        self.attn_hidden_units = attn_hidden_units
+        self.vocab_size = vocab_size
+        self.embedding_dim = embedding_dim
+
+        self.attention = BahdanauAttention(attn_hidden_units)
+
+        # Convert word indices to embedding vectors.
+        self.embedding = tf.keras.layers.Embedding(input_dim=vocab_size,
+                                                   output_dim=embedding_dim)
+
+        self.lstm = tf.keras.layers.LSTM(units=self.hidden_units,
+                                         return_sequences=True,
+                                         return_state=True,
+                                         recurrent_initializer='glorot_uniform')
+
+        self.dense1 = tf.keras.layers.Dense(units=hidden_units)
+
+        # Final layer to convert into "scores" per each vocabulary.
+        self.dense2 = tf.keras.layers.Dense(units=vocab_size)
+
+    # pylint: disable=arguments-differ
+    def call(
+            self, prev_word: tf.Tensor, features: tf.Tensor,
+            prev_hidden_state: tf.Tensor
+    ) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
+        """
+        Arguments:
+            prev_word: Sequence of indices of previous words,
+                used as ingredients to predict the next `word_seq_len` words.
+                `shape == (batch, word_seq_len)`.
```

Comment:
Since we are embedding the input vocab into the embedding dimension, `word_seq_len` should actually be `vocab_size`.
Also, since we are sampling one word given only 1 previous word, the shape should actually be `(batch, 1, vocab_size)`.
This changes part of your forward call.

Unless you were assuming that `word_seq_len == 1`, in which case the shape should be `(batch, word_seq_len, vocab_size`

Reply:
Shit, I misunderstood the `input_dim` parameter of `tf.Kerala.layers.Embedding` to be the input’s channel size.
Yer right!

Reply:
Oh true. The api is indeed a bit confusing.